cpp/src/arrow/compute/exec/project_benchmark.cc

```diff
@@ -0,0 +1,174 @@
+#include "arrow/dataset/partition.h"
+#include "arrow/testing/future_util.h"
+#include "arrow/testing/generator.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/type.h"
+
+namespace arrow {
+namespace compute {
+
+static constexpr int64_t total_batch_size = 1e6;
+
+static void ProjectionOverhead(benchmark::State& state, Expression expr) {
+  const auto batch_size = static_cast<int32_t>(state.range(0));
+  const auto num_batches = total_batch_size / batch_size;
+
+  auto data = MakeRandomBatches(schema({field("i64", int64()), field("bool", boolean())}),
+                                num_batches, batch_size);
+  ExecContext ctx(default_memory_pool(), arrow::internal::GetCpuThreadPool());
+  for (auto _ : state) {
+    state.PauseTiming();
+    ASSERT_OK_AND_ASSIGN(auto plan, ExecPlan::Make(&ctx));
+    AsyncGenerator<util::optional<ExecBatch>> sink_gen;
+    ASSERT_OK(Declaration::Sequence(
+                  {
+                      {"source",
+                       SourceNodeOptions{data.schema,
+                                         data.gen(/*parallel=*/true, /*slow=*/false)},
+                       "custom_source_label"},
+                      {"project", ProjectNodeOptions{{
+                                      expr,
+                                  }}},
+                      {"sink", SinkNodeOptions{&sink_gen}, "custom_sink_label"},
+                  })
+                  .AddToPlan(plan.get()));
+    state.ResumeTiming();
+    ASSERT_FINISHES_OK(StartAndCollect(plan.get(), sink_gen));
+  }
+  state.counters["rows_per_second"] = benchmark::Counter(
+      static_cast<double>(state.iterations() * num_batches * batch_size),
+      benchmark::Counter::kIsRate);
+
+  state.counters["batches_per_second"] = benchmark::Counter(
+      static_cast<double>(state.iterations() * num_batches), benchmark::Counter::kIsRate);
+}
+
+static void ProjectionOverheadIsolated(benchmark::State& state, Expression expr) {
+  const auto batch_size = static_cast<int32_t>(state.range(0));
+  const auto num_batches = total_batch_size / batch_size;
+
+  auto data = MakeRandomBatches(schema({field("i64", int64()), field("bool", boolean())}),
+                                num_batches, batch_size);
+  ExecContext ctx(default_memory_pool(), arrow::internal::GetCpuThreadPool());
+  for (auto _ : state) {
+    state.PauseTiming();
+
+    AsyncGenerator<util::optional<ExecBatch>> sink_gen;
+    ASSERT_OK_AND_ASSIGN(auto plan, ExecPlan::Make(&ctx));
+    // Source and sink nodes have no effect on the benchmark.
+    // Used for dummy purposes as they are referenced in InputReceived and InputFinished.
+    ASSERT_OK_AND_ASSIGN(
+        auto sn, MakeExecNode("source", plan.get(), {},
+                              SourceNodeOptions{data.schema, data.gen(/*parallel=*/true,
+                                                                      /*slow=*/false)}));
+    ASSERT_OK_AND_ASSIGN(auto pn, MakeExecNode("project", plan.get(), {sn},
```

Comment:
Minor nit: Let's try and avoid overly short variable names (e.g. `sn`, `pn`, `tp`).  I suppose this is somewhat hypocritical of me since I used `cv` but even that could be something like `all_tasks_finished_cv`.

Reference: https://google.github.io/styleguide/cppguide.html#Constant_Names